### ISSUE TITLE
Support ed25519 keys in KeyRing

### ISSUE THREAD

**Issue by kevinji** [informalsystems/hermes#2851]
## Summary

Add [ed25519](https://ed25519.cr.yp.to/) support to the relayer's KeyRing.

## Problem Definition

I'm looking to add Solana compatibility to the relayer, which requires adding support for ed25519 keys instead of secp256k1 keys. Ideally, KeyRing is generic over the key type as long as it's defined by a trait.

## Proposal

Modify `KeyRing` to instead be `KeyRing<T: Curve>` and define `impl Curve for Secp256k1` and `impl Curve for Ed25519`. Add support for edd25519 using the [curve25519-dalek crate](https://github.com/dalek-cryptography/curve25519-dalek). This might also require `AddressType` to include a new `Solana` address type, and other interface changes that I'm not 100% sure of yet.

## Acceptance Criteria

KeyRing supports both secp256k1 and ed25519 keys.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by romac** [comment]
Hi, thanks for bringing this to our attention! Are you currently working on this?

**Comment by kevinji** [comment]
@romac Yep I'm planning to work on this if there's support for making this change.

**Comment by romac** [comment]
Totally! Would be great if you could post a rough draft of the design you have in mind, and then we can greenlight it :)

**Comment by kevinji** [comment]
Sure, here's a summary of what I think the high-level changes would look like:
- Add a type parameter `T` to [`keyring::KeyEntry`](https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/keyring.rs) where `T: Curve`, `Curve` being a new trait we define that encapsulates public and private keys for both secp256k1 and ed25519.
  - Modify `public_key` and `private_key` in `KeyEntry` to refer to types defined in `T`. (De)serialization should remain the same for secp256k1. Possibly also update `account` although it's not as meaningful as the pubkey is just the account address for Solana.
- Implement [`signature::Signer`](https://docs.rs/signature/latest/signature/trait.Signer.html) and [`signature::Verifier`](https://docs.rs/signature/latest/signature/trait.Verifier.html) for both secp256k1 and ed25519, to allow sharing some code.
- Add a new [`config::AddressType`](https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/config.rs) called `Ed25519`.
- Modify `KeyFile` so that public and private keys can be stored directly for ed25519 instead of using Bech32 and a mnemonic.

I'll probably run into more things as I finish more of the ed25519 changes.

**Comment by romac** [comment]
Sounds great! Let us if we can help with anything, and otherwise we'll make to review your PR in a timely manner and will do our best to include in 1.2 provided there's enough time for that (planned release date is Dec 11th at the moment).
